### PR TITLE
Validation for scene and caption IDs

### PR DESCRIPTION
**Pull request by kehahn**

### CODE REVIEW

File: src/server/cea_608_encoder/byte_pair_generator.py [hunk 1]
```diff
@@ -76,6 +77,10 @@ def consume_scenes(scene_list: list) -> list:
 
         if 'scene_id' not in scene:
             raise ValueError('Every scene must have a scene ID.')
+        else if scene['scene_id'] in scene_id_check:
```
Comment (Markobozic) [inline src/server/cea_608_encoder/byte_pair_generator.py]: This is a good start but it's very inefficient as you have to scan this `scene_id_check` list potentially tons of times if we have lots of scenes.

What if we create a function `validate_scene_ids(scene_list)` that loops through each scene and adds scene_id along with how many times it has been seen into a dictionary: `dict = {'scene_id': 'times_it_has_been_seen'}` (this is just pseudo code) and then when we are done adding them, we loop through the dictionary looking for a `times_it_has_been_seen` that is > 1. We can also use this information to check which `scene_names` are conflicting and return that to the user for it to be easier to debug.

Comment (kehahn) [reply src/server/cea_608_encoder/byte_pair_generator.py]: Ok, I wrote validation functions instead. Let me know what you think.

File: src/server/cea_608_encoder/byte_pair_generator.py [hunk 2]
```diff
@@ -167,3 +173,32 @@ def consume_captions(caption_list: list) -> dict:
...
+
+
+def validate_scene_ids(scene_list: list)
+    """Validates the scene IDs to look for duplicate IDs
+
+    :param scene_list:
+    """
+    scene_ids = {}
+    for scene in scene_list:
+        scene_ids.update(scene['scene_id'] += 1)
```
Comment (Markobozic) [inline src/server/cea_608_encoder/byte_pair_generator.py]: is this line working for you? I get this error trying to run it:
```
 def validate_scene_ids(scene_list: list)
                                           ^
SyntaxError: invalid syntax
```

File: src/server/cea_608_encoder/byte_pair_generator.py [hunk 3]
```diff
@@ -67,7 +67,8 @@ def consume_scenes(scene_list: list) -> list:
     :return: TODO
     """
     scene_data = []
-    
+    validate_scene_ids(scene_list)
```
Comment (Markobozic) [inline src/server/cea_608_encoder/byte_pair_generator.py]: This call occurs prior to checking whether there currently exists a scene id in a scene. This line: `if 'scene_id' not in scene:` is where we check whether the id even exists but if we call your validation before then we can get a KeyError when you try `scene['scene_id']`

File: src/server/cea_608_encoder/byte_pair_generator.py (hunk 2 above)
Comment (kehahn) [reply src/server/cea_608_encoder/byte_pair_generator.py]: Was missing a colon...

File: src/server/cea_608_encoder/byte_pair_generator.py (hunk 3 above)
Comment (kehahn) [reply src/server/cea_608_encoder/byte_pair_generator.py]: Changed it to be called afterwards

**Review by Markobozic — COMMENTED** [review]
The code doesn't run when i input valid caption data; i get `TypeError: unhashable type: 'dict'`

File: src/server/cea_608_encoder/byte_pair_generator.py [hunk 4]
```diff
@@ -67,14 +68,16 @@ def consume_scenes(scene_list: list) -> list:
...
+
     for scene in scene_list:
         current_scene_data = {}
         current_scene_data['start'] = 0
         current_scene_data['data'] = []
 
         if 'scene_id' not in scene:
             raise ValueError('Every scene must have a scene ID.')
+        else:
+            validate_scene_ids(scene_list)
```
Comment (Markobozic) [inline src/server/cea_608_encoder/byte_pair_generator.py]: This and the caption id validation call are happening within for-loops so for each scene were gonna validate every single scene id. If we have 100 scenes, we will perform the same validation 100 times ^^

Comment (kehahn) [reply src/server/cea_608_encoder/byte_pair_generator.py]: Where do you recommend I put it?

Comment (Markobozic) [reply src/server/cea_608_encoder/byte_pair_generator.py]: Probably at the end of the for-loop, outside of it.

File: src/server/cea_608_encoder/byte_pair_generator.py [hunk 5]
```diff
@@ -166,4 +173,31 @@ def consume_captions(caption_list: list) -> dict:
...
+def validate_scene_ids(scene_list: list):
+    """Validates the scene IDs to look for duplicate IDs
+
+    :param scene_list:
+    """
+    scene_ids = Counter(scene_list)
+
+    for id,number_of_that_id in scene_ids.items():
+        if number_of_that_id > 1:
+            raise ValueError('There are duplicate scene IDs.')
```
Comment (Markobozic) [inline src/server/cea_608_encoder/byte_pair_generator.py]: if i had 200 scenes and got this error back, how would i go about finding which two scenes are conflicting?

File: src/server/cea_608_encoder/byte_pair_generator.py [hunk 6]
```diff
@@ -132,7 +134,11 @@ def consume_captions(caption_list: list) -> dict:
     caption_metadata = {}
     caption_metadata['caption_string'] = []
 
+
     for caption in caption_list:
+        if 'caption_id' not in caption or 'caption_string' not in caption:
```
Comment (Markobozic) [inline src/server/cea_608_encoder/byte_pair_generator.py]: The check for caption id and caption string should already exists, this looks like a duplicate check

Comment (kehahn) [reply src/server/cea_608_encoder/byte_pair_generator.py]: This was something that was already here before I started working on PER-397, but I can remove this.

**Review by Markobozic — APPROVED** [review]
Looks awesome, great job!

**Pull request merged**